File: Hard/Better String/better-string.java
//{ Driver Code Starts
// Initial Template for Java

import java.util.*;
import java.lang.*;
import java.math.*;
import java.io.*;

class GFG {
    public static void main(String[] args) throws IOException {
        Scanner sc = new Scanner(System.in);
        int t = sc.nextInt();
        while (t-- > 0) {
            String str1 = sc.next();
            String str2 = sc.next();

            Solution obj = new Solution();
            String ans = obj.betterString(str1, str2);
            System.out.println(ans);
        }
    }
}
// } Driver Code Ends


// User function Template for Java

class Solution {
    public static String betterString(String str1, String str2) {
        // Code here
        int count1 = countDistinctSubsequences(str1);
        int count2 = countDistinctSubsequences(str2);
        return (count1 >= count2) ? str1 : str2;
    }
    private static int countDistinctSubsequences(String str) {
        int MOD = 1000000007;
        int n = str.length();
        int[] lastOccurrence = new int[256];
        Arrays.fill(lastOccurrence, -1);
        int[] dp = new int[n + 1];
        dp[0] = 1; 
        for (int i = 1; i <= n; i++) {
            dp[i] = (2 * dp[i - 1]) % MOD;
            if (lastOccurrence[str.charAt(i - 1)] != -1) {
                dp[i] = (dp[i] - dp[lastOccurrence[str.charAt(i - 1)] - 1] + MOD) % MOD;
            }
            lastOccurrence[str.charAt(i - 1)] = i;
        }
        return dp[n];
    }
}

File: Hard/Better String/better-string.cpp
//{ Driver Code Starts
// Initial Template for C++

#include <bits/stdc++.h>
using namespace std;


// } Driver Code Ends
// User function Template for C++

class Solution {
  public:
    string betterString(string str1, string str2) {
        // code here
         int a = countSubSeq(str1);
    int b = countSubSeq(str2);
    if (a >= b) {
        return str1;
    } else {
        return str2;
    }
}

int countSubSeq(string &str) {
    int n = str.size();
    int dp[n + 1] = {};
    int mod = 1e9 + 7;

    // Initialize with an empty subsequence
    dp[0] = 1;

    // Initialize the last position of each character
    vector<int> lastPos(26, -1);

    for (int i = 1; i <= n; i++) {
        dp[i] = (2 * dp[i - 1]) % mod;

        // If the character has occurred before, subtract the count of subsequences
        if (lastPos[str[i - 1] - 'a'] != -1) {
            dp[i] = (dp[i] - dp[lastPos[str[i - 1] - 'a'] - 1] + mod) % mod;
        }

        // Update the last position of the current character
        lastPos[str[i - 1] - 'a'] = i;
    }

    return (dp[n] - 1 + mod) % mod;  // Subtract 1 to get the count of distinct subsequences
}
};

//{ Driver Code Starts.

int main() {
    int t;
    cin >> t;
    while (t--) {
        string str1, str2;
        cin >> str1 >> str2;
        Solution obj;
        string ans = obj.betterString(str1, str2);
        cout << ans << "\n";
    }
}

// } Driver Code Ends